libzim/writer.py

```diff
@@ -143,43 +184,68 @@ def __exit__(self, *args):
         self._creatorWrapper.add_article(article)
         if not article.is_redirect():
-            self._update_article_counter(article)
-
-    def _update_article_counter(self, article):
-        # default dict update
-        self._article_counter[article.get_mime_type().strip()] += 1
-
-    def mandatory_metadata_ok(self):
-        """Flag if mandatory metadata is complete and not empty"""
-        metadata_item_ok = [k in self._metadata for k in MANDATORY_METADATA_KEYS]
-        return all(metadata_item_ok)
-
-    def update_metadata(self, **kwargs):
-        "Updates article metadata" ""
-        new_metadata = {pascalize(k): v for k, v in kwargs.items()}
-        self._metadata.update(new_metadata)
-
-    def write_metadata(self):
-        for key, value in self._metadata.items():
-            if key == "Date" and isinstance(value, datetime.date):
-                value = value.strftime("%Y-%m-%d")
-            article = MetadataArticle(key, value)
-            self._creatorWrapper.add_article(article)
+            # update article counter
+            self._article_counter[article.get_mime_type().strip()] += 1
 
-        article = MetadataArticle("Counter", self._get_counter_string())
-        self._creatorWrapper.add_article(article)
+    @property
+    def mandatory_metadata_ok(self) -> bool:
+        """ Whether all mandatory metadata have been provided and not empty """
+        return all([k in self._metadata for k in MANDATORY_METADATA_KEYS])
+
+    def update_metadata(self, **kwargs: str):
+        """ Updates Creator metadata for ZIM, supplied as keyword arguments """
+
+        def pascalize(keyword: str):
+            """ Converts python case to pascal case.
+
+                example: long_description -> LongDescription """
+            return "".join(keyword.title().split("_"))
 
-    def _get_counter_string(self):
-        return ";".join(["%s=%s" % (k, v) for (k, v) in self._article_counter.items()])
+        self._metadata.update({pascalize(k): v for k, v in kwargs.items()})
 
     def close(self):
-        self.write_metadata()
+        """ Finalizes and writes added articles to the file
+
+            Raises
+            ------
+                RuntimeError
+                    If the ZimCreator was already finalized """
+
+        def get_counter_string() -> str:
+            return ";".join([f"{k}={v}" for (k, v) in self._article_counter.items()])
+
+        def write_metadata():
+            """Store _medtadata dict as MetadataArticle"""
+            for key, value in self._metadata.items():
+                if key == "Date" and isinstance(value, (datetime.date, datetime.datetime)):
+                    value = value.strftime("%Y-%m-%d")
+                article = MetadataArticle(key, value)
+                self._creatorWrapper.add_article(article)
+
+            article = MetadataArticle("Counter", get_counter_string())
+            self._creatorWrapper.add_article(article)
```

Comment:
I would not create a function just to call it after.
Simply write the code in the `close` body.

Reply:
works for me